controllers/metal3datatemplate_controller.go

```diff
@@ -0,0 +1,176 @@
+// +kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile handles Metal3Machine events
+func (r *Metal3DataTemplateReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx := context.Background()
+	metadataLog := r.Log.WithName(dataTemplateControllerName).WithValues("metal3-datatemplate", req.NamespacedName)
+
+	// Fetch the Metal3DataTemplate instance.
+	capm3Metadata := &capm3.Metal3DataTemplate{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, capm3Metadata); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	helper, err := patch.NewHelper(capm3Metadata, r.Client)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to init patch helper")
+	}
+	// Always patch capm3Machine exiting this function so we can persist any Metal3Machine changes.
+	defer func() {
+		err := helper.Patch(ctx, capm3Metadata)
+		if err != nil {
+			metadataLog.Info("failed to Patch capm3Metadata")
+		}
+	}()
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, capm3Metadata.ObjectMeta)
+	if capm3Metadata.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err != nil {
+			metadataLog.Info("Metal3DataTemplate is missing cluster label or cluster does not exist, Requeuing")
+			return ctrl.Result{}, nil
+		}
+		if cluster == nil {
+			metadataLog.Info(fmt.Sprintf("This metadata is not yet associated with a cluster using the label %s: <name of cluster>", capi.ClusterLabelName))
+			return ctrl.Result{}, nil
+		}
+	}
+
+	if cluster != nil {
+		metadataLog = metadataLog.WithValues("cluster", cluster.Name)
+
+		// Return early if the Metadata or Cluster is paused.
+		if util.IsPaused(cluster, capm3Metadata) {
+			metadataLog.Info("reconciliation is paused for this object")
+			return ctrl.Result{Requeue: true, RequeueAfter: requeueAfter}, nil
+		}
+	}
+
+	// Create a helper for managing the metadata object.
+	metadataMgr, err := r.ManagerFactory.NewDataTemplateManager(capm3Metadata, metadataLog)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create helper for managing the metadata")
+	}
+
+	// Handle deleted metadata
+	if !capm3Metadata.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, metadataMgr)
+	}
+
+	// Handle non-deleted machines
+	return r.reconcileNormal(ctx, metadataMgr)
+}
+
+func (r *Metal3DataTemplateReconciler) reconcileNormal(ctx context.Context,
+	metadataMgr baremetal.DataTemplateManagerInterface,
+) (ctrl.Result, error) {
+	// If the Metal3DataTemplate doesn't have finalizer, add it.
+	metadataMgr.SetFinalizer()
+
+	err := metadataMgr.RecreateStatusConditionally(ctx)
+	if err != nil {
+		return checkMetadataError(err, "Failed to recreate the status")
```

Comment:
what is meant by recreate status?

Reply:
that is if the object was moved, and the status is gone, it will repopulate it